File: solutions/go/pov/1/pov.go
package pov

import "slices"

type Tree struct {
    size		int
	value		string
    children	[]*Tree
    parent		*Tree
}

func New(value string, children ...*Tree) *Tree {
    size := 1
    tr := &Tree{0, value, children, nil}
    for _, child := range children {
        size += child.size
        child.parent = tr
    }
    tr.size = size
    return tr
}

func (tr *Tree) Value() string {
	return tr.value
}

func (tr *Tree) Children() []*Tree {
	return tr.children
}

func (tr *Tree) String() string {
	if tr == nil {
		return "nil"
	}
	result := tr.Value()
	if len(tr.Children()) == 0 {
		return result
	}
	for _, ch := range tr.Children() {
		result += " " + ch.String()
	}
	return "(" + result + ")"
}

func (tr *Tree) getPath(to string) []*Tree {
    if tr == nil || tr.size == 0 {
        return nil
    }
    dfs := make([]*Tree, 0, tr.size)
    path := make([]*Tree, 0, tr.size)
    dfs = append(dfs, tr)
    var node *Tree
    var found bool
    for len(dfs) > 0 {
        node = dfs[len(dfs) - 1]
        dfs = dfs[:len(dfs) - 1]
        if node.value == to {
            found = true
            break
        }
        for _, child := range node.children {
            dfs = append(dfs, child)
        }
    }
    if !found {
        return nil
    }

    for node != nil {
        path = append(path, node)
        node = node.parent
    }
    slices.Clip(path)
    slices.Reverse(path)
    return path
}

func (tr *Tree) swap(child *Tree) {
    tr.size, child.size = tr.size - child.size, tr.size
    tr.parent, child.parent = child, tr.parent
    child.children = append(child.children, tr)
    tr.children = slices.DeleteFunc(tr.children, func(t *Tree) bool {
        return t == child
    })
}

func (tr *Tree) FromPov(from string) *Tree {
	path := tr.getPath(from)
    if path == nil || len(path) == 0 {
        return nil
    }
    if len(path) == 1 {
        return tr
    }
    for i, node := range path[:len(path) - 1] {
        node.swap(path[i + 1])
    }
    return path[len(path) - 1]
}

func (tr *Tree) PathTo(from, to string) []string {
	newTree := tr.FromPov(from)
    if newTree == nil {
        return []string{}
    }
    path := newTree.getPath(to)
    if path == nil || len(path) == 0 {
        return []string{}
    }
    result := make([]string, len(path))
    for i := range result {
        result[i] = path[i].value
    }
    return result
}

File: solutions/python/pov/1/pov.py
from json import dumps

class Tree:
    def __init__(self, label, children=None):
        self.label = label
        self.children = children if children is not None else []

    def __dict__(self):
        return {self.label: [c.__dict__() for c in sorted(
            self.children, key = lambda ch: ch.label
        )]}

    def __hash__(self):
        return id(self)

    def __str__(self, indent=None):
        return dumps(self.__dict__(), indent=indent)

    def __lt__(self, other):
        return self.label < other.label

    def __eq__(self, other):
        return self.__dict__() == other.__dict__()

    def matches(self, name):
        return self.label == name
    
    def index(self, child):
        for i, ch in enumerate(self.children):
            if ch == child:
                return i
        return -1

    def add(self, child):
        self.children.append(child)

    def remove(self, child):
        cursor = self.index(child)
        if cursor >= 0:
            return self.children.pop(cursor)
        return None

    def swap(self, parent):
        if not parent.remove(self):
            return None
        self.add(parent)
        return parent

    def swap_to_root(self, trail):
        cursor = self
        while cursor in trail:
            parent = trail[cursor]
            cursor = cursor.swap(parent)
        return self

    def from_pov(self, from_node):
        trail = {}
        stack = [self]
        while stack:
            cursor = stack.pop()
            if cursor.matches(from_node):
                return cursor.swap_to_root(trail)
            for child in cursor.children:
                trail[child] = cursor
                stack.append(child)
        raise ValueError('Tree could not be reoriented')

    def ancestry(self, node_name):
        trail = {}
        stack = [self]
        while stack:
            cursor = stack.pop()
            if cursor.matches(node_name):
                ancestry_list = []
                while cursor is not None:
                    ancestry_list.append(cursor.label)
                    cursor = trail.get(cursor)
                return ancestry_list
            for child in cursor.children:
                trail[child] = cursor
                stack.append(child)
        return None

    def path_to(self, from_node, to_node):
        from_src = self.ancestry(from_node)
        from_dst = self.ancestry(to_node)
        if from_src is None:
            raise ValueError('Tree could not be reoriented')
        if from_dst is None:
            raise ValueError('No path found')
        last_removed = None
        while from_src and from_dst and from_src[-1] == from_dst[-1]:
            last_removed = from_src.pop()
            from_dst.pop()
        if last_removed is not None:
            from_src.append(last_removed)
        from_src.extend(from_dst[::-1])
        return from_src